### DISCUSSION TITLE
Policy to check if email exists in array

### DISCUSSION THREAD

**Discussion by praj18** [supabase/supabase#4511] [Questions]
*_I have a schema similar to that of reddit's so I have changed the table names and columns to make it easier to understand_

I am having issues creating a policy for an update function which involves an array.

I've got 3 tables subreddits, posts and comments which are visually represented as follows.

### subreddits
|  subreddit_id  | moderators | ... |
|:----:|:--------------:|:---:|
| 1    | [test@test.com, elonmusk@tsla.com]   | ... |
| 2    | [mark@zucc.com]   | ... |

### posts
|  subreddit_id  | post_id | ... |
|:----:|:--------------:|:---:|
| 1    | 18   | ... |
| 2    | 35   | ... |

### comments
|  id  |  post_id  | comment_text | author | status |
|:----:|:----:|:--------------:|----------:|----------:|
| 1 | 18    | 'Hello world'   |81283. | active |
| 2 | 18   | 'comment 2'   | 75563 | flagged |
| 3 | 35   | 'comment 123'   | 71143 | deleted |

When selecting moderators for the subreddit in which the comment exists, I can run the following query.

```
SELECT moderators FROM subreddit
JOIN posts
 ON subreddit.subreddit_id = posts.subreddit_id
INNER JOIN comments
ON comments.post_id = posts.id
 WHERE comments.id = 1;
```

And the following line returns the subreddit IDs where the current user is a moderator.

```
select subreddit_id from subreddits where 'test@test.com'=ANY(moderators);
```

### TLDR

When a comment is posted by a user on to a post in a particular subreddit, I intend to allow the moderators to edit the comment's status and also the author of the comment to delete the comment (however, deleting doesn't actually delete the comment, but rather changes the status to deleted).

Now, my question is, how do I formulate a policy that only allows updates if the current user who is logged in is either the moderator of the subreddit (has an email that exists in the moderator array) or the author of the comment.

**Comment by tchief** [comment]
Try create a function
```
create or replace function can_user_update_comment(user_id int, user_email text, comment_id int) returns boolean as $$
    SELECT (
        EXISTS (
            SELECT 1 FROM comments WHERE id=$3 AND author=$1
        )
        OR EXISTS (
            SELECT moderators FROM subreddit
            JOIN posts ON subreddit.subreddit_id = posts.subreddit_id
            INNER JOIN comments ON comments.post_id = posts.id
            WHERE comments.id = $3 AND $2 = ANY(moderators)
        )
    )
$$ stable language sql security definer;
```
and then use it in policy
```
can_user_update_comment(auth.uid(), auth.email(), id)
```
